### ISSUE TITLE
„Cannot create Launcher without at least one TestEngine“ — When there is indeed a TestEngine

### ISSUE THREAD

**Issue by lingocoder** [gradle/gradle#11373]
### Expected Behavior

When the Java Gradle Plugin is applied in a project, it should find test framework launcher classes and service providers that are present on the *`testRuntimeClasspath`*.

### Current Behavior

When the Java Gradle Plugin is applied in a project, it *fails* to find test framework launcher classes and service providers that are absolutely, most definitely, and provably present on the *`:testRuntimeClasspath`*.

### Context

Test tasks that are configured to execute unit tests that cover classes of a JPMS module, work perfectly fine in a project that does not have the Java Gradle Plugin applied.

Doing nothing more than merely applying the Java Gradle Plugin in a module project, however, causes JUnit 5 test execution to fail with...

```
...
Caused by: org.gradle.internal.impldep.org.junit.platform.commons.util.PreconditionViolationException: Cannot create Launcher without at least one TestEngine; consider adding an engine implementation JAR to the classpath
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.platform.commons.util.Preconditions.condition(Preconditions.java:295)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.platform.launcher.core.DefaultLauncher.<init>(DefaultLauncher.java:58)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.platform.launcher.core.LauncherFactory.create(LauncherFactory.java:91)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.platform.launcher.core.LauncherFactory.create(LauncherFactory.java:67)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:97)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 25 more
```

...even though the JUnit 5 dependencies are configured and can be clearly observed to be present on the *`testRuntimeClasspath`*...

```
...
:test's classpath ---

   '    ...:$GRADLE_USER_HOME/caches/modules-2/files-2.1/org.junit.jupiter/junit-jupiter-engine/5.5.2/.../junit-jupiter-engine-5.5.2.jar:$GRADLE_USER_HOME/caches/modules-2/files-2.1/org.junit.jupiter/junit-jupiter-api/5.5.2/.../junit-jupiter-api-5.5.2.jar:$GRADLE_USER_HOME/caches/modules-2/files-2.1/org.junit.platform/junit-platform-launcher/1.5.2/.../junit-platform-launcher-1.5.2.jar:$GRADLE_USER_HOME/caches/modules-2/files-2.1/org.junit.platform/junit-platform-engine/1.5.2/.../junit-platform-engine-1.5.2.jar:$GRADLE_USER_HOME/caches/modules-2/files-2.1/org.junit.platform/junit-platform-commons/1.5.2/.../junit-platform-commons-1.5.2.jar:$GRADLE_USER_HOME/caches/modules-2/files-2.1/org.apiguardian/apiguardian-api/1.1.0/.../apiguardian-api-1.1.0.jar:$GRADLE_USER_HOME/caches/modules-2/files-2.1/org.opentest4j/opentest4j/1.2.0/.../opentest4j-1.2.0.jar    '
	
...
```

Those exact same JUnit 5 JARs are also in the module path. As proposed by JUnit 5 & OpenJDK devs @marcphilipp & @sormuras, having JUnit 5 on both paths is [*considered to be current best practice*](https://github.com/openjdk/skara/pull/66#issuecomment-523874046).

JUnit 4 tests, fail under the same conditions. But with a different error message...

```
...
java.lang.Exception: No runnable methods
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.runners.BlockJUnit4ClassRunner.validateInstanceMethods(BlockJUnit4ClassRunner.java:191)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.runners.BlockJUnit4ClassRunner.collectInitializationErrors(BlockJUnit4ClassRunner.java:128)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.runners.ParentRunner.validate(ParentRunner.java:416)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.runners.ParentRunner.<init>(ParentRunner.java:84)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:65)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.internal.builders.JUnit4Builder.runnerForClass(JUnit4Builder.java:10)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at gradle.api@6.0-rc-2/org.gradle.internal.impldep.org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:33)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:78)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at gradle.api@6.0-rc-2/org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	...
	at gradle.api@6.0-rc-2/org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at gradle.api@6.0-rc-2/org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at gradle.api@6.0-rc-2/org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at gradle.api@6.0-rc-2/org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
    ...
```

### Steps to Reproduce 

1. Download [*this demonstration project*](http://bit.ly/antiJUnit5)

2. In a terminal, *`cd`* into the root folder of the project and run *`:check`* 

3. Observe in the generated HTML report, the errors reported above

4. Observe in the stdout printed to the terminal, that *all* of the JUnit 5 JARs — including the *`org.junit.platform.launcher`* module — are indeed present in the classpath and in the module path (*[as recommended](https://github.com/openjdk/skara/pull/66#issuecomment-523874046)*)

5. In the project's *`build.gradle`* script, comment out line 14 so that the build does not use the Java Gradle Plugin:

    *`id 'java-gradle-plugin'`*

6. Run *`:check`* again.

7. Observe that the test is successfully launched and passes

### Your Environment

```
JUnit 5 v5.5.2
Gradle 6.0-rc-2
OpenJDK 12
Windows amd64
```

**Comment by marcphilipp** [comment]
@lingocoder Thanks for raising the issue!

I'm getting the following error when running your example project with Gradle 6.0:
https://scans.gradle.com/s/efpan3mgfefak/console-log?task=:test#L26

Any ideas?

**Comment by lingocoder** [comment]
Hi @marcphilipp,

Thanks for looking into this.

> „*Any ideas?*“

Yeah. In a nutshell, that error is the symptom of a JPMS split package. The root cause of the split package [*was reported in this jigsaw-dev mailing list msg*](http://bit.ly/jigsawDev014310)

> ...
> *For some reason, Gradle's generated artifact redundantly duplicates all*
> *the packages of two modules from the JDK: javax.xml and jdk.xml.dom.*
> ...

To which one of the original authors of the Jigsaw/JPMS specification replied:

> *The Gradle devs will know but I assume these are coming from the* 
> *xml-apis-${version}.jar (Apache Xerces project). It's not clear, to me* 
> *anyway, why these API classes are still being distributed in 2019. The* 
> *XML processing API was added to Java SE 1.4, the XML stream API came a* 
> *bit later. The XML APIs started out in Java SE as "standalone* 
> *technologies" where they were in the JDK but could be upgraded by* 
> *deploying a newer version with the endorsed standards override* 
> *mechanism. It's possible, and probable, that developers have been* 
> *deploying xml-apis on the class path and got lucky that it didn't cause* 
> *problems. Anyway, none of this makes sense since Java SE 9 as both* 
> *JSR-206 and JSR-173 ceased to be standalone technologies (see the final* 
> *MR for both JSRs), and mechanism to upgrade them (the endorsed standards* 
> *override mechanism) has been removed.*

I was actually in the process — *right before seeing your reply  a moment ago* — of nearly completing an implementation of my plugin that works around that split package issue.

Of course my workaround is merely a band-aid to unblock my progress — in the short term. But ideally, Gradle devs would hopefully be able to implement a better (*or as as @AlanBateman recommends in the mailing list, a* „*saner*“) permanent solution to the root problem.

I appreciate you probably can't make any more progress past the blocker than I did when I reported this. But what would you recommend yourself, marcphilipp? Going on what you've been able to tell so far?

TIA :+1:

**Comment by marcphilipp** [comment]
Could you please raise a separate issue to remove `javax.xml.transform` from the gradle-api JAR?

**Comment by lingocoder** [comment]
> „*Could you please raise a separate issue to remove `javax.xml.transform` from the gradle-api JAR?*“


Would that it were that straightforward, marcphilipp :)

Here is just a random sample of the internal Gradle implementation dependencies a removal of that package's containing module (*`java.xml`*) would impact:

```gradle-api-6.0-rc-2.jar
   org.gradle.internal.impldep.org.mozilla.javascript.xmlimpl -> javax.xml.transform                                gradle-api-6.0-rc-2.jar
   org.gradle.internal.impldep.org.mozilla.javascript.xmlimpl -> javax.xml.transform.dom                            gradle-api-6.0-rc-2.jar
   org.gradle.internal.impldep.org.mozilla.javascript.xmlimpl -> javax.xml.transform.stream                         gradle-api-6.0-rc-2.jar
   ...
org.gradle.internal.xml                            -> javax.xml.transform                                gradle-api-6.0-rc-2.jar
   org.gradle.internal.xml                            -> javax.xml.transform.dom                            gradle-api-6.0-rc-2.jar
   org.gradle.internal.xml                            -> javax.xml.transform.stream                         gradle-api-6.0-rc-2.jar
...   
```

And *`javax.xml.transform`* is just one package of several packages in the *`java.xml`* module. Even if you removed the entire *`java.xml`* module somehow,  There's a second module (*`jdk.xml.dom`*) in Gradle's shaded JAR (*itself with a ton more packages*) that you'd have to contend with. It also causes split packages in the exact same way. 

There are tons more classes that would be impacted. So many, that you would very likely be annoyed at me if I pasted them all here. But [*this might give you a flavor*](http://bit.ly/xmlapisHits) of the general impact those two modules have within the Gradle code base. 

You could run the *`jdeps`* command on the *`gradle-api-6.0.jar`* to get a more detailed picture of the impact if you need more.

The really frustrating thing that I've discovered after seeking advice in the jigsaw-dev mailing list, is that it's impossible to patch the jar the way you would ordinarily solve a split package dilemma.

I'm curious, marcphilipp. Were you able to reproduce any part of the issue I described in my original report? I gather you can't share any specific solution that might have eluded me so far?

In any case, I sincerely appreciate that you took the time to look into this. Many thanks :+1:

**Comment by marcphilipp** [comment]
> Were you able to reproduce any part of the issue I described in my original report?

No, I ran into the issue I linked above.

Could you please elaborate a bit more on your use case. Why do you want to run tests for a Gradle plugin using the module path? It will not be executed that way at runtime so I don't see the point.

**Comment by lingocoder** [comment]
Thanks marcphilipp,

I explain the use case in my typical boringly-verbose style [*here*](https://github.com/java9-modularity/gradle-modules-plugin/issues/117).

***Tl;DR*** — I'm developing a Gradle plugin that enables JPMS development. Among other JPMS development-related capabilities, it also turns a project into a JPMS modular Multi-Release JAR File. It does that — *not only for non-plugin projects* — also for projects that are themselves plugins.

> „...*It will not be executed that way at runtime so I don't see the point*...“

It's not a plugin *run-time* issue. But a plugin *development-time* issue. The issue is only encountered when my plugin is doing its thing on another plugin project. Because only plugin development brings in *`gradle-api-${version}.jar`*.

I discovered the issue after recently refactoring my plugin, and applying that newly-refactored version to that same *`gradle-modules-plugin`* project I linked to above. I use that project as a sort of functional test for the *plugin-as-a-modular-MRJAR* use case. 

I do concede it is a rare use case. But  [***mrJar***](http://bit.ly/mrJar) provides that capability, nevertheless, Think of it as an *experimental* or *incubating* feature — *to use Gradle's terminology*

**Comment by sormuras** [comment]
> Tl;DR — I'm developing a Gradle plugin that enables JPMS development. Among other JPMS development-related capabilities, it also turns a project into a JPMS modular Multi-Release JAR File. It does that — not only for non-plugin projects — also for projects that are themselves plugins.

+1 ... _Modules, modules, everywhere!_

**Comment by marcphilipp** [comment]
Why would a Gradle plugin project need to be a "JPMS modular Multi-Release JAR"?

**Comment by lingocoder** [comment]
> „*Why would a Gradle plugin project need to be a "JPMS modular Multi-Release JAR"?*“

Thanks for asking marcphilipp

Here's _one_ explanation from the link I shared previously:

>> ...
>>
>>_...Depending on how a particular plugin is designed, the capabilities it provides to a Gradle build, might not be physically coded in the plugin's implementations of *`Plugin`* and *`DefaulTask`*. Those implementations are merely entry points to the *actual functionality* of the plugin._ 
>> 
>> _That *actual functionality* could be implemented by *behind-the-scenes* collaborators to which the *`Plugin`* and *`DefaultTask`* implementations delegate. Couldn't there be some value in explicitly exporting that *`behind-the-scenes`* functionality to be available for use outside of the context of a Gradle build? Obviously not in your plugin — because it wasn't *designed to be a library*. But could there be some value in doing it in other plugin projects?_
>>
>> _If there *could be*, then there *could be* some value in making a plugin's distributable, usable outside of the inflexible constraints of *Gradle World*. If a plugin's JAR were also an explicit module, its actual *behind-the-scenes* implementation details could additionally be useful in the wider *Java World* outside of a Gradle build. That is one distinguishing *point* of ***mrJar***._
>>
>> ...

**Comment by marcphilipp** [comment]
In that case, I'd propose extracting that behind-the-scenes functionality into a separate subproject that is consumed by the gradle plugin project.
